### ISSUE TITLE
RuntimeException: Internal error in Firestore (20.1.0).

### ISSUE THREAD

**Issue by TomislavMladenov** [firebase/firebase-android-sdk#596]
**Describe your environment**

  * Android Studio version: 3.4.2
  * Firebase Component: FireStore
  * Component version: 20.1.0

 **Describe the problem**

Just started implementing FireStore and while I was trying to get the dummy data from my firestore I got this error. Every time I try to start the app it crashes. If I remove the firestore dependency everything is working fine. So after 2 days trying to solve it searching about it, I simple wasn't able to do so. If you can help in any way will be appreciated

**Steps to reproduce:**

Start the app.

** Relevant Code:**

_Gradle app_

```gradle
apply plugin: 'com.android.application'

android {
    compileSdkVersion 29
    buildToolsVersion "29.0.0"
    defaultConfig {
        applicationId "com.example.sieredelivery"
        minSdkVersion 22
        targetSdkVersion 29
        multiDexEnabled true
        versionCode 1
        versionName "1.0"
        testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
    }
    buildTypes {
        release {
            minifyEnabled false
            proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
        }
    }

    dataBinding {
        enabled = true
    }

    android {
        compileOptions {
            sourceCompatibility JavaVersion.VERSION_1_8
            targetCompatibility JavaVersion.VERSION_1_8
        }
    }
}

dependencies {
    //Support libraries
    implementation fileTree(dir: 'libs', include: ['*.jar'])
    implementation 'androidx.appcompat:appcompat:1.0.2'
    implementation 'androidx.constraintlayout:constraintlayout:1.1.3'
    implementation 'androidx.cardview:cardview:1.0.0'
    implementation 'androidx.recyclerview:recyclerview:1.0.0'
    implementation 'androidx.viewpager:viewpager:1.0.0'
    implementation 'com.google.android.material:material:1.1.0-alpha07'
    implementation 'com.google.android.material:material:1.0.0'
    //Firebase
    implementation 'com.google.firebase:firebase-auth:18.0.0'
    implementation 'com.google.android.gms:play-services-auth:17.0.0'
    implementation 'com.google.firebase:firebase-core:17.0.0'
    implementation 'com.google.gms:google-services:4.3.0'
    implementation 'com.google.firebase:firebase-storage:18.0.0'
    implementation 'com.google.firebase:firebase-database:18.0.0'
    implementation 'com.google.firebase:firebase-functions:18.0.0'
    implementation 'com.google.firebase:firebase-messaging:19.0.1'
    implementation 'com.google.firebase:firebase-perf:18.0.1'
    implementation 'com.firebase:firebase-client-android:2.3.1'
    implementation 'com.google.firebase:firebase-firestore:20.1.0'
    implementation 'androidx.multidex:multidex:2.0.1'
    //Room
    implementation "androidx.room:room-runtime:2.1.0"
    annotationProcessor "androidx.room:room-compiler:2.1.0"

    //test
    testImplementation 'junit:junit:4.12'
    androidTestImplementation 'androidx.test:runner:1.2.0'
    androidTestImplementation 'androidx.test.espresso:espresso-core:3.2.0'
}

apply plugin: 'com.google.gms.google-services'
```

_Gradle project_

```gradle
buildscript {
    repositories {
        google()
        jcenter()

        
    }
    dependencies {
        classpath 'com.android.tools.build:gradle:3.4.2'
        classpath 'com.google.gms:google-services:4.2.0'


        // NOTE: Do not place your application dependencies here; they belong
        // in the individual module build.gradle files
    }
}

allprojects {
    repositories {
        google()
        jcenter()
        
    }
}

task clean(type: Delete) {
    delete rootProject.buildDir
}
```

_MainActivity (test run on the fireStore)_
```java
 @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main);
        setPresenter();
        setToolbar();
        setBottomNavigation();
        loadFragment(SearchFragment.getInstance());
        createNewStore();


    }

    private void createNewStore() {
        FirebaseFirestore db = FirebaseFirestore.getInstance();

        db.collection("stores")
                .get()
                .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<QuerySnapshot> task) {
                        if (task.isSuccessful()) {
                            for (QueryDocumentSnapshot document : task.getResult()) {
                                Log.d(TAG, document.getId() + " => " + document.getData());
                            }
                        } else {
                            Log.w(TAG, "Error getting documents.", task.getException());
                        }
                    }
                });
    }
```

**Error**

```
java.lang.RuntimeException: Internal error in Firestore (20.1.0).
        at com.google.firebase.firestore.util.AsyncQueue.lambda$panic$5(com.google.firebase:firebase-firestore@@20.1.0:379)
        at com.google.firebase.firestore.util.AsyncQueue$$Lambda$5.run(com.google.firebase:firebase-firestore@@20.1.0)
        at android.os.Handler.handleCallback(Handler.java:815)
        at android.os.Handler.dispatchMessage(Handler.java:104)
        at android.os.Looper.loop(Looper.java:194)
        at android.app.ActivityThread.main(ActivityThread.java:5643)
        at java.lang.reflect.Method.invoke(Native Method)
        at java.lang.reflect.Method.invoke(Method.java:372)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:960)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
     Caused by: java.lang.NoClassDefFoundError: Failed resolution of: Lcom/google/firebase/firestore/local/SQLiteSchema$$Lambda$6;
        at com.google.firebase.firestore.local.SQLiteSchema.ensureSequenceNumbers(com.google.firebase:firebase-firestore@@20.1.0:364)
        at com.google.firebase.firestore.local.SQLiteSchema.runMigrations(com.google.firebase:firebase-firestore@@20.1.0:123)
        at com.google.firebase.firestore.local.SQLiteSchema.runMigrations(com.google.firebase:firebase-firestore@@20.1.0:74)
        at com.google.firebase.firestore.local.SQLitePersistence$OpenHelper.onCreate(com.google.firebase:firebase-firestore@@20.1.0:302)
        at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:251)
        at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:163)
        at com.google.firebase.firestore.local.SQLitePersistence.start(com.google.firebase:firebase-firestore@@20.1.0:125)
        at com.google.firebase.firestore.core.FirestoreClient.initialize(com.google.firebase:firebase-firestore@@20.1.0:254)
        at com.google.firebase.firestore.core.FirestoreClient.lambda$new$2(com.google.firebase:firebase-firestore@@20.1.0:117)
        at com.google.firebase.firestore.core.FirestoreClient$$Lambda$2.run(com.google.firebase:firebase-firestore@@20.1.0)
        at com.google.firebase.firestore.util.AsyncQueue.lambda$enqueue$4(com.google.firebase:firebase-firestore@@20.1.0:311)
        at com.google.firebase.firestore.util.AsyncQueue$$Lambda$4.call(com.google.firebase:firebase-firestore@@20.1.0)
        at com.google.firebase.firestore.util.AsyncQueue.lambda$enqueue$3(com.google.firebase:firebase-firestore@@20.1.0:287)
        at com.google.firebase.firestore.util.AsyncQueue$$Lambda$3.run(com.google.firebase:firebase-firestore@@20.1.0)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:422)
        at java.util.concurrent.FutureTask.run(FutureTask.java:237)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:152)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:265)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
        at com.google.firebase.firestore.util.AsyncQueue$DelayedStartFactory.run(com.google.firebase:firebase-firestore@@20.1.0:205)
        at java.lang.Thread.run(Thread.java:818)
     Caused by: java.lang.ClassNotFoundException: Didn't find class "com.google.firebase.firestore.local.SQLiteSchema$$Lambda$6" on path: DexPathList[[zip file "/data/app/com.example.sieredelivery-1/base.apk"],nativeLibraryDirectories=[/vendor/lib64, /system/lib64]]
        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:511)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:469)
        at com.google.firebase.firestore.local.SQLiteSchema.ensureSequenceNumbers(com.google.firebase:firebase-firestore@@20.1.0:364) 
        at com.google.firebase.firestore.local.SQLiteSchema.runMigrations(com.google.firebase:firebase-firestore@@20.1.0:123) 
        at com.google.firebase.firestore.local.SQLiteSchema.runMigrations(com.google.firebase:firebase-firestore@@20.1.0:74) 
        at com.google.firebase.firestore.local.SQLitePersistence$OpenHelper.onCreate(com.google.firebase:firebase-firestore@@20.1.0:302) 
        at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:251) 
        at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:163) 
        at com.google.firebase.firestore.local.SQLitePersistence.start(com.google.firebase:firebase-firestore@@20.1.0:125) 
        at com.google.firebase.firestore.core.FirestoreClient.initialize(com.google.firebase:firebase-firestore@@20.1.0:254) 
        at com.google.firebase.firestore.core.FirestoreClient.lambda$new$2(com.google.firebase:firebase-firestore@@20.1.0:117) 
        at com.google.firebase.firestore.core.FirestoreClient$$Lambda$2.run(com.google.firebase:firebase-firestore@@20.1.0) 
        at com.google.firebase.firestore.util.AsyncQueue.lambda$enqueue$4(com.google.firebase:firebase-firestore@@20.1.0:311) 
        at com.google.firebase.firestore.util.AsyncQueue$$Lambda$4.call(com.google.firebase:firebase-firestore@@20.1.0) 
        at com.google.firebase.firestore.util.AsyncQueue.lambda$enqueue$3(com.google.firebase:firebase-firestore@@20.1.0:287) 
        at com.google.firebase.firestore.util.AsyncQueue$$Lambda$3.run(com.google.firebase:firebase-firestore@@20.1.0) 
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:422) 
        at java.util.concurrent.FutureTask.run(FutureTask.java:237) 
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:152) 
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:265) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587) 
        at com.google.firebase.firestore.util.AsyncQueue$DelayedStartFactory.run(com.google.firebase:firebase-firestore@@20.1.0:205) 
        at java.lang.Thread.run(Thread.java:818) 
    	Suppressed: java.lang.NoClassDefFoundError: com.google.firebase.firestore.local.SQLiteSchema$$Lambda$6
        at dalvik.system.DexFile.defineClassNative(Native Method)
        at dalvik.system.DexFile.defineClass(DexFile.java:226)
        at dalvik.system.DexFile.loadClassBinaryName(DexFile.java:219)
    		at dalvik.system.DexPat
```

**Comment by h-amg** [comment]
@TomislavMladenov I have the exact same problem, here is what i noticed while trying to debug it:

-  It occurs only after retrieving data from firestore while writing data works fine.
- The data is actually retrieved successfully (it shows up in the log).
- The app crashes well after data retrieval.

**Comment by var-const** [comment]
@h-amg Could you please provide some additional information?
* the exact exception message;
* debug logs (use `FirebaseFirestore.setLoggingEnabled(true)` to enable logging; please make sure the logs contain no sensitive information).

It is surprising that you would experience the exception described in the original post while also have logs that show successful data retrieval. The original exception occurs in a part of code that runs _before_ any other Firestore operation can run, so it would appear no data retrieval should be possible.

**Comment by var-const** [comment]
@TomislavMladenov Thanks for the repro case, I will try it out locally and see if I can reproduce the issue.

**Comment by h-amg** [comment]
@var-const seems like i got it all wrong.

1. My exception message is different to the one in the post:
> Caused by: java.lang.VerifyError: Verifier rejected class com.google.firestore.v1.WriteResponse due to bad method java.lang.Object com.google.firestore.v1.WriteResponse.dynamicMethod(com.google.protobuf.GeneratedMessageLite$MethodToInvoke, java.lang.Object, java.lang.Object) (declaration of 'com.google.firestore.v1.WriteResponse' appears in /data/app/com.app.android.name-1/split_lib_dependencies_apk.apk:classes2.dex)

2. Data retrieval was actually unsuccessful, code runs normally anywhere but inside firestore onComplete callback method:
      try{
            userDocRef = db.collection("***s").document("******************sa");
            userDocRef.get().addOnCompleteListener(new OnCompleteListener<DocumentSnapshot>() {
                @Override
                public void onComplete(@NonNull Task<DocumentSnapshot> task) {
                    if (task.isSuccessful()) {
                        DocumentSnapshot document = task.getResult();
                        if (document.exists()) {
                            Log.d(TAG, "DocumentSnapshot data: " + document.getData());
                        } else {
                            Log.d(TAG, "No such document");
                        }
                    } else {
                        Log.d(TAG, "get failed with ", task.getException());
                    }
                }
            });
        }catch (Exception e){
            Log.d(TAG, "Error retrieving user doc data: " + e);
        }


**DebugLogs:**

> 07-11 08:02:11.603 27666-27666/com.app.android.name W/DisplayListCanvas: DisplayListCanvas is started on unbinded RenderNode (without mOwningView)
> 07-11 08:02:11.613 27666-27666/com.app.android.name D/ViewRootImpl: MSG_RESIZED_REPORT: ci=Rect(0, 72 - 0, 0) vi=Rect(0, 72 - 0, 0) or=1
> 07-11 08:02:11.653 27666-28056/com.app.android.name I/DynamiteModule: Considering local module providerinstaller:0 and remote module providerinstaller:0
> 07-11 08:02:11.653 27666-28056/com.app.android.name W/ProviderInstaller: Failed to load providerinstaller module: No acceptable module found. Local version is 0 and remote version is 0.
> 07-11 08:02:11.653 27666-28056/com.app.android.name W/ResourcesManager: getTopLevelResources: /data/app/com.google.android.gms-1/base.apk / 1.0 running in com.app.android.name rsrc of package null
> 07-11 08:02:11.683 27666-27666/com.app.android.name I/Timeline: Timeline: Activity_idle id: android.os.BinderProxy@ff29e8d time:147037202
> 07-11 08:02:11.723 27666-29453/com.app.android.name I/art: Verification error in void com.google.firestore.v1.WriteResponse.mergeCommitTime(com.google.protobuf.Timestamp)
> 07-11 08:02:11.723 27666-29453/com.app.android.name I/art: void com.google.firestore.v1.WriteResponse.mergeCommitTime(com.google.protobuf.Timestamp): [0x10] couldn't find method com.google.protobuf.Timestamp$Builder.mergeFrom (Lcom/google/protobuf/GeneratedMessageLite;)Lcom/google/protobuf/GeneratedMessageLite$Builder;
> 07-11 08:02:11.723 27666-29453/com.app.android.name I/art: void com.google.firestore.v1.WriteResponse.mergeCommitTime(com.google.protobuf.Timestamp) failed to verify: void com.google.firestore.v1.WriteResponse.mergeCommitTime(com.google.protobuf.Timestamp): [0x10] register v3 has type Precise Reference: com.google.protobuf.Timestamp but expected Reference: com.google.protobuf.GeneratedMessageLite
> 07-11 08:02:11.723 27666-29453/com.app.android.name I/art: Verification error in java.lang.Object com.google.firestore.v1.WriteResponse.dynamicMethod(com.google.protobuf.GeneratedMessageLite$MethodToInvoke, java.lang.Object, java.lang.Object)
> 07-11 08:02:11.723 27666-29453/com.app.android.name I/art: java.lang.Object com.google.firestore.v1.WriteResponse.dynamicMethod(com.google.protobuf.GeneratedMessageLite$MethodToInvoke, java.lang.Object, java.lang.Object): [0x5C] couldn't find method com.google.protobuf.Timestamp.toBuilder ()Lcom/google/protobuf/GeneratedMessageLite$Builder;
> 07-11 08:02:11.723 27666-29453/com.app.android.name I/art: java.lang.Object com.google.firestore.v1.WriteResponse.dynamicMethod(com.google.protobuf.GeneratedMessageLite$MethodToInvoke, java.lang.Object, java.lang.Object): [0x73] couldn't find method com.google.protobuf.Timestamp$Builder.mergeFrom (Lcom/google/protobuf/GeneratedMessageLite;)Lcom/google/protobuf/GeneratedMessageLite$Builder;
> 07-11 08:02:11.723 27666-29453/com.app.android.name I/art: java.lang.Object com.google.firestore.v1.WriteResponse.dynamicMethod(com.google.protobuf.GeneratedMessageLite$MethodToInvoke, java.lang.Object, java.lang.Object) failed to verify: java.lang.Object com.google.firestore.v1.WriteResponse.dynamicMethod(com.google.protobuf.GeneratedMessageLite$MethodToInvoke, java.lang.Object, java.lang.Object): [0x73] register v5 has type Precise Reference: com.google.protobuf.Timestamp but expected Reference: com.google.protobuf.GeneratedMessageLite
> 07-11 08:02:11.723 27666-29453/com.app.android.name W/art: Verification failed on class com.google.firestore.v1.WriteResponse in /data/app/com.app.android.name-1/split_lib_dependencies_apk.apk because: Verifier rejected class com.google.firestore.v1.WriteResponse due to bad method java.lang.Object com.google.firestore.v1.WriteResponse.dynamicMethod(com.google.protobuf.GeneratedMessageLite$MethodToInvoke, java.lang.Object, java.lang.Object)
> 07-11 08:02:11.733 27666-27666/com.app.android.name D/AndroidRuntime: Shutting down VM
> 07-11 08:02:11.733 27666-27666/com.app.android.name E/AndroidRuntime: FATAL EXCEPTION: main
>     Process: com.app.android.name, PID: 27666
>     java.lang.RuntimeException: Internal error in Firestore (20.1.0).
>         at com.google.firebase.firestore.util.AsyncQueue.lambda$panic$5(com.google.firebase:firebase-firestore@@20.1.0:379)
>         at com.google.firebase.firestore.util.AsyncQueue$$Lambda$5.run(com.google.firebase:firebase-firestore@@20.1.0)
>         at android.os.Handler.handleCallback(Handler.java:739)
>         at android.os.Handler.dispatchMessage(Handler.java:95)
>         at android.os.Looper.loop(Looper.java:158)
>         at android.app.ActivityThread.main(ActivityThread.java:7224)
>         at java.lang.reflect.Method.invoke(Native Method)
>         at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1230)
>         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1120)
>      Caused by: java.lang.VerifyError: Verifier rejected class com.google.firestore.v1.WriteResponse due to bad method java.lang.Object com.google.firestore.v1.WriteResponse.dynamicMethod(com.google.protobuf.GeneratedMessageLite$MethodToInvoke, java.lang.Object, java.lang.Object) (declaration of 'com.google.firestore.v1.WriteResponse' appears in /data/app/com.app.android.name-1/split_lib_dependencies_apk.apk)
>         at com.google.firestore.v1.WriteResponse.getDefaultInstance(com.google.firebase:firebase-firestore@@20.1.0:1012)
>         at com.google.firestore.v1.FirestoreGrpc.getWriteMethod(com.google.firebase:firebase-firestore@@20.1.0:379)
>         at com.google.firebase.firestore.remote.WriteStream.<init>(com.google.firebase:firebase-firestore@@20.1.0:74)
>         at com.google.firebase.firestore.remote.Datastore.createWriteStream(com.google.firebase:firebase-firestore@@20.1.0:104)
>         at com.google.firebase.firestore.remote.RemoteStore.<init>(com.google.firebase:firebase-firestore@@20.1.0:186)
>         at com.google.firebase.firestore.core.FirestoreClient.initialize(com.google.firebase:firebase-firestore@@20.1.0:263)
>         at com.google.firebase.firestore.core.FirestoreClient.lambda$new$2(com.google.firebase:firebase-firestore@@20.1.0:117)
>         at com.google.firebase.firestore.core.FirestoreClient$$Lambda$2.run(com.google.firebase:firebase-firestore@@20.1.0)
>         at com.google.firebase.firestore.util.AsyncQueue.lambda$enqueue$4(com.google.firebase:firebase-firestore@@20.1.0:311)
>         at com.google.firebase.firestore.util.AsyncQueue$$Lambda$4.call(com.google.firebase:firebase-firestore@@20.1.0)
>         at com.google.firebase.firestore.util.AsyncQueue.lambda$enqueue$3(com.google.firebase:firebase-firestore@@20.1.0:287)
>         at com.google.firebase.firestore.util.AsyncQueue$$Lambda$3.run(com.google.firebase:firebase-firestore@@20.1.0)
>         at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:423)
>         at java.util.concurrent.FutureTask.run(FutureTask.java:237)
>         at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:154)
>         at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:269)
>         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
>         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
>         at com.google.firebase.firestore.util.AsyncQueue$DelayedStartFactory.run(com.google.firebase:firebase-firestore@@20.1.0:205)
>         at java.lang.Thread.run(Thread.java:818)
> 07-11 08:02:11.863 27666-27685/com.app.android.name W/art: Suspending all threads took: 62.326ms
> 07-11 08:02:11.893 27666-27685/com.app.android.name I/art: Background sticky concurrent mark sweep GC freed 78605(5MB) AllocSpace objects, 48(2MB) LOS objects, 18% free, 28MB/35MB, paused 63.394ms total 143.944ms
> 07-11 08:02:11.963 27666-28056/com.app.android.name V/NativeCrypto: Registering com/google/android/gms/org/conscrypt/NativeCrypto's 284 native methods...
> 07-11 08:02:12.073 27666-28056/com.app.android.name I/art: Rejecting re-init on previously-failed class java.lang.Class<com.google.android.gms.org.conscrypt.Java7ExtendedSSLSession>
> 07-11 08:02:12.083 27666-28056/com.app.android.name I/art: Rejecting re-init on previously-failed class java.lang.Class<com.google.android.gms.org.conscrypt.Java7ExtendedSSLSession>
> 07-11 08:02:12.083 27666-28056/com.app.android.name I/art: Rejecting re-init on previously-failed class java.lang.Class<com.google.android.gms.org.conscrypt.Java8ExtendedSSLSession>
> 07-11 08:02:12.083 27666-28056/com.app.android.name I/art: Rejecting re-init on previously-failed class java.lang.Class<com.google.android.gms.org.conscrypt.Java8ExtendedSSLSession>
> 07-11 08:02:12.133 27666-28056/com.app.android.name I/ProviderInstaller: Installed default security provider GmsCore_OpenSSL

**Note:**
I had duplicate classes issue between firebase and dialogflow earlier and i excluded them using the following in build.gradle file:
       android {
       
           ...
       
           configurations {
               implementation.exclude module: 'protolite-well-known-types'
               implementation.exclude module: 'protobuf-lite' /*implementation.exclude module: 'protobuf-java'*/
           }
       }

**Comment by TomislavMladenov** [comment]
@var-const  I can send you the source code or upload it to github if neeeded, to see if it will help reproduce it. I just started the app so there's basically there is nothing just basic things.

**Comment by var-const** [comment]
@TomislavMladenov It would be super-helpful! You can upload it to Github or send it to my email, `varconst` at `google.com`.  Please make sure there's no sensitive data before sharing.

**Comment by var-const** [comment]
@h-amg It looks like a different issue. I will take a look; in the meantime, can you please open a new issue in this repo and just copy all the info you provided there?

**Comment by var-const** [comment]
@TomislavMladenov Update: I couldn't reproduce the issue. I created a new app from scratch and used the code/Gradle snippets you provided. I had to add a tiny bit of logic to add a document to the collection, but after that, the document is read successfully. So I would need a repro case to continue with the issue.

**Comment by pranaytokopedia** [comment]
i am also facing the same issue

**Comment by wilhuff** [comment]
At this point, where we are with this issue is that we're not clear on what could be causing the problem but can't reproduce it so we have no way to investigate. If you're also facing this issue, then the best way to help us track it down is to help us reproduce the problem, as @var-const [requested above](https://github.com/firebase/firebase-android-sdk/issues/596#issuecomment-510638899).

There are a few ways you can do this:

  * Package up your project and share it with us
  * Put together detailed instructions for how we can recreate the issue

As it stands there's not a whole lot we can do to help.

**Comment by pranaytokopedia** [comment]
from what i have investigated, this seems to be only reproduced in android 5.x and 6.x devices. so it maybe because these android versions dont support java lamdas oe something like that. i am using firestore version 17.1.2 and im getting the same error , just that the name of the generated lamda class is different
`Caused by: java.lang.NoClassDefFoundError: Failed resolution of: Lcom/google/firebase/firestore/local/SQLiteSchema$$Lambda$2;`

**Comment by BlueRayNL** [comment]
No , I am having the same issue on android Q version 9+

**Comment by rylexr** [comment]
Some users of my app are facing same issue. Here's the data I got:

```
Fatal Exception: java.lang.RuntimeException: Internal error in Firestore (20.2.0).
       at com.google.firebase.firestore.util.AsyncQueue.lambda$panic$5 + 379(AsyncQueue.java:379)
       at com.google.firebase.firestore.util.AsyncQueue$$Lambda$5.run(AsyncQueue.java)
       at android.os.Handler.handleCallback + 815(Handler.java:815)
       at android.os.Handler.dispatchMessage + 104(Handler.java:104)
       at android.os.Looper.loop + 207(Looper.java:207)
       at android.app.ActivityThread.main + 5728(ActivityThread.java:5728)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run + 789(ZygoteInit.java:789)
       at com.android.internal.os.ZygoteInit.main + 679(ZygoteInit.java:679)
Caused by java.lang.NoClassDefFoundError: Failed resolution of: Lcom/google/firebase/firestore/L/Ga;
       at com.google.firebase.firestore.local.SQLiteSchema.ensureSequenceNumbers + 364(SQLiteSchema.java:364)
       at com.google.firebase.firestore.local.SQLiteSchema.runMigrations + 123(SQLiteSchema.java:123)
       at com.google.firebase.firestore.local.SQLiteSchema.runMigrations + 74(SQLiteSchema.java:74)
       at com.google.firebase.firestore.local.SQLitePersistence$OpenHelper.onCreate + 302(SQLitePersistence.java:302)
       at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked + 251(SQLiteOpenHelper.java:251)
       at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase + 163(SQLiteOpenHelper.java:163)
       at com.google.firebase.firestore.local.SQLitePersistence.start + 125(SQLitePersistence.java:125)
       at com.google.firebase.firestore.core.FirestoreClient.initialize + 254(FirestoreClient.java:254)
       at com.google.firebase.firestore.core.FirestoreClient.lambda$new$2 + 117(FirestoreClient.java:117)
       at com.google.firebase.firestore.core.FirestoreClient$$Lambda$2.run(FirestoreClient.java)
       at com.google.firebase.firestore.util.AsyncQueue.lambda$enqueue$4 + 311(AsyncQueue.java:311)
       at com.google.firebase.firestore.util.AsyncQueue$$Lambda$4.call(AsyncQueue.java)
       at com.google.firebase.firestore.util.AsyncQueue.lambda$enqueue$3 + 287(AsyncQueue.java:287)
       at com.google.firebase.firestore.util.AsyncQueue$$Lambda$3.run(AsyncQueue.java)
       at java.util.concurrent.Executors$RunnableAdapter.call + 423(Executors.java:423)
       at java.util.concurrent.FutureTask.run + 237(FutureTask.java:237)
       at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run + 269(ScheduledThreadPoolExecutor.java:269)
       at java.util.concurrent.ThreadPoolExecutor.runWorker + 1113(ThreadPoolExecutor.java:1113)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run + 588(ThreadPoolExecutor.java:588)
       at com.google.firebase.firestore.util.AsyncQueue$DelayedStartFactory.run + 205(AsyncQueue.java:205)
       at java.lang.Thread.run + 818(Thread.java:818)
Caused by java.lang.ClassNotFoundException: Didn't find class "com.google.firebase.firestore.L.Ga" on path: DexPathList[[zip file "/data/app/[APP ID REMOVED]-1/base.apk", zip file "/data/app/[APP ID REMOVED]-1/split_config.armeabi_v7a.apk", zip file "/data/app/[APP ID REMOVED]-1/split_config.en.apk", zip file "/data/app/[APP ID REMOVED]-1/split_config.xhdpi.apk"],nativeLibraryDirectories=[/data/app/[APP ID REMOVED]-1/lib/arm, /data/app/[APP ID REMOVED]-1/base.apk!/lib/armeabi-v7a, /data/app/[APP ID REMOVED]-1/split_config.armeabi_v7a.apk!/lib/armeabi-v7a, /data/app/[APP ID REMOVED]-1/split_config.en.apk!/lib/armeabi-v7a, /data/app/[APP ID REMOVED]-1/split_config.xhdpi.apk!/lib/armeabi-v7a, /vendor/lib, /system/lib]]
       at dalvik.system.BaseDexClassLoader.findClass + 56(BaseDexClassLoader.java:56)
       at java.lang.ClassLoader.loadClass + 511(ClassLoader.java:511)
       at java.lang.ClassLoader.loadClass + 469(ClassLoader.java:469)
       at com.google.firebase.firestore.local.SQLiteSchema.ensureSequenceNumbers + 364(SQLiteSchema.java:364)
       at com.google.firebase.firestore.local.SQLiteSchema.runMigrations + 123(SQLiteSchema.java:123)
       at com.google.firebase.firestore.local.SQLiteSchema.runMigrations + 74(SQLiteSchema.java:74)
       at com.google.firebase.firestore.local.SQLitePersistence$OpenHelper.onCreate + 302(SQLitePersistence.java:302)
       at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked + 251(SQLiteOpenHelper.java:251)
       at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase + 163(SQLiteOpenHelper.java:163)
       at com.google.firebase.firestore.local.SQLitePersistence.start + 125(SQLitePersistence.java:125)
       at com.google.firebase.firestore.core.FirestoreClient.initialize + 254(FirestoreClient.java:254)
       at com.google.firebase.firestore.core.FirestoreClient.lambda$new$2 + 117(FirestoreClient.java:117)
       at com.google.firebase.firestore.core.FirestoreClient$$Lambda$2.run(FirestoreClient.java)
       at com.google.firebase.firestore.util.AsyncQueue.lambda$enqueue$4 + 311(AsyncQueue.java:311)
       at com.google.firebase.firestore.util.AsyncQueue$$Lambda$4.call(AsyncQueue.java)
       at com.google.firebase.firestore.util.AsyncQueue.lambda$enqueue$3 + 287(AsyncQueue.java:287)
       at com.google.firebase.firestore.util.AsyncQueue$$Lambda$3.run(AsyncQueue.java)
       at java.util.concurrent.Executors$RunnableAdapter.call + 423(Executors.java:423)
       at java.util.concurrent.FutureTask.run + 237(FutureTask.java:237)
       at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run + 269(ScheduledThreadPoolExecutor.java:269)
       at java.util.concurrent.ThreadPoolExecutor.runWorker + 1113(ThreadPoolExecutor.java:1113)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run + 588(ThreadPoolExecutor.java:588)
       at com.google.firebase.firestore.util.AsyncQueue$DelayedStartFactory.run + 205(AsyncQueue.java:205)
       at java.lang.Thread.run + 818(Thread.java:818)
```

**Devices showing this issue**
Brand: samsung
Model: Galaxy A5(2017)
OS: Version: 6.0.1

Brand: LENOVO
Model: VIBE K10
OS: Version: 6.0

Brand: Sony
Model: Xperia X
OS: Version: 6.0.1

Brand: samsung
Model: Galaxy Note3
OS: Version: 5.0

Brand: Meizu
Model: m3 note
OS: Version: 5.1

Brand: samsung
Model: Galaxy On7 Pro
OS: Version: 6.0.1

Brand: LENOVO
Model: K3 Note
OS: Version: 6.0

None of these devices are rooted. Is there a way you can reopen this issue? Currently, this is # 1 issue on my Crashlytics report and it's affecting a lot of users. I'll be happy to provide more details if you need.

Regards!

**Comment by schmidt-sebastian** [comment]
@rylexr - Is there a chance you are pulling in two different version of the Firestore SDK? Early releases of our SDK were proguarded, which causes long class names to be rewritten to much shorter names. It looks like part of your stacktrace is Proguarded ("com/google/firebase/firestore/L/Ga"), but other places are not ("com.google.firebase.firestore.local.SQLiteSchema.ensureSequenceNumbers").

Can you verify the dependency resolution of your build?

**Comment by rylexr** [comment]
I have a multi-module app and Firestore SDK was included (same version) in both modules. I've refactored code to include Firestore SDK just in one module. I'll report back if issue persist.

**Comment by rylexr** [comment]
Unfortunately, issue still persist. Here's stacktrace:

```
Fatal Exception: java.lang.RuntimeException: Internal error in Firestore (20.2.0).
       at com.google.firebase.firestore.util.AsyncQueue.lambda$panic$5 + 379(AsyncQueue.java:379)
       at com.google.firebase.firestore.util.AsyncQueue$$Lambda$5.run(AsyncQueue.java)
       at android.os.Handler.handleCallback + 815(Handler.java:815)
       at android.os.Handler.dispatchMessage + 104(Handler.java:104)
       at android.os.Looper.loop + 207(Looper.java:207)
       at android.app.ActivityThread.main + 5728(ActivityThread.java:5728)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run + 789(ZygoteInit.java:789)
       at com.android.internal.os.ZygoteInit.main + 679(ZygoteInit.java:679)
Caused by java.lang.NoClassDefFoundError: Failed resolution of: Lcom/google/firebase/firestore/L/Ga;
       at com.google.firebase.firestore.local.SQLiteSchema.ensureSequenceNumbers + 364(SQLiteSchema.java:364)
       at com.google.firebase.firestore.local.SQLiteSchema.runMigrations + 123(SQLiteSchema.java:123)
       at com.google.firebase.firestore.local.SQLiteSchema.runMigrations + 74(SQLiteSchema.java:74)
       at com.google.firebase.firestore.local.SQLitePersistence$OpenHelper.onCreate + 302(SQLitePersistence.java:302)
       at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked + 251(SQLiteOpenHelper.java:251)
       at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase + 163(SQLiteOpenHelper.java:163)
       at com.google.firebase.firestore.local.SQLitePersistence.start + 125(SQLitePersistence.java:125)
       at com.google.firebase.firestore.core.FirestoreClient.initialize + 254(FirestoreClient.java:254)
       at com.google.firebase.firestore.core.FirestoreClient.lambda$new$2 + 117(FirestoreClient.java:117)
       at com.google.firebase.firestore.core.FirestoreClient$$Lambda$2.run(FirestoreClient.java)
       at com.google.firebase.firestore.util.AsyncQueue.lambda$enqueue$4 + 311(AsyncQueue.java:311)
       at com.google.firebase.firestore.util.AsyncQueue$$Lambda$4.call(AsyncQueue.java)
       at com.google.firebase.firestore.util.AsyncQueue.lambda$enqueue$3 + 287(AsyncQueue.java:287)
       at com.google.firebase.firestore.util.AsyncQueue$$Lambda$3.run(AsyncQueue.java)
       at java.util.concurrent.Executors$RunnableAdapter.call + 423(Executors.java:423)
       at java.util.concurrent.FutureTask.run + 237(FutureTask.java:237)
       at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run + 269(ScheduledThreadPoolExecutor.java:269)
       at java.util.concurrent.ThreadPoolExecutor.runWorker + 1113(ThreadPoolExecutor.java:1113)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run + 588(ThreadPoolExecutor.java:588)
       at com.google.firebase.firestore.util.AsyncQueue$DelayedStartFactory.run + 205(AsyncQueue.java:205)
       at java.lang.Thread.run + 818(Thread.java:818)
Caused by java.lang.ClassNotFoundException: Didn't find class "com.google.firebase.firestore.L.Ga" on path: DexPathList[[zip file "/data/app/[APP ID REMOVED]-1/base.apk", zip file "/data/app/[APP ID REMOVED]-1/split_config.armeabi_v7a.apk", zip file "/data/app/[APP ID REMOVED]-1/split_config.en.apk", zip file "/data/app/[APP ID REMOVED]-1/split_config.xhdpi.apk"],nativeLibraryDirectories=[/data/app/[APP ID REMOVED]-1/lib/arm, /data/app/[APP ID REMOVED]-1/base.apk!/lib/armeabi-v7a, /data/app/[APP ID REMOVED]-1/split_config.armeabi_v7a.apk!/lib/armeabi-v7a, /data/app/[APP ID REMOVED]-1/split_config.en.apk!/lib/armeabi-v7a, /data/app/[APP ID REMOVED]-1/split_config.xhdpi.apk!/lib/armeabi-v7a, /vendor/lib, /system/lib]]
       at dalvik.system.BaseDexClassLoader.findClass + 56(BaseDexClassLoader.java:56)
       at java.lang.ClassLoader.loadClass + 511(ClassLoader.java:511)
       at java.lang.ClassLoader.loadClass + 469(ClassLoader.java:469)
       at com.google.firebase.firestore.local.SQLiteSchema.ensureSequenceNumbers + 364(SQLiteSchema.java:364)
       at com.google.firebase.firestore.local.SQLiteSchema.runMigrations + 123(SQLiteSchema.java:123)
       at com.google.firebase.firestore.local.SQLiteSchema.runMigrations + 74(SQLiteSchema.java:74)
       at com.google.firebase.firestore.local.SQLitePersistence$OpenHelper.onCreate + 302(SQLitePersistence.java:302)
       at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked + 251(SQLiteOpenHelper.java:251)
       at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase + 163(SQLiteOpenHelper.java:163)
       at com.google.firebase.firestore.local.SQLitePersistence.start + 125(SQLitePersistence.java:125)
       at com.google.firebase.firestore.core.FirestoreClient.initialize + 254(FirestoreClient.java:254)
       at com.google.firebase.firestore.core.FirestoreClient.lambda$new$2 + 117(FirestoreClient.java:117)
       at com.google.firebase.firestore.core.FirestoreClient$$Lambda$2.run(FirestoreClient.java)
       at com.google.firebase.firestore.util.AsyncQueue.lambda$enqueue$4 + 311(AsyncQueue.java:311)
       at com.google.firebase.firestore.util.AsyncQueue$$Lambda$4.call(AsyncQueue.java)
       at com.google.firebase.firestore.util.AsyncQueue.lambda$enqueue$3 + 287(AsyncQueue.java:287)
       at com.google.firebase.firestore.util.AsyncQueue$$Lambda$3.run(AsyncQueue.java)
       at java.util.concurrent.Executors$RunnableAdapter.call + 423(Executors.java:423)
       at java.util.concurrent.FutureTask.run + 237(FutureTask.java:237)
       at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run + 269(ScheduledThreadPoolExecutor.java:269)
       at java.util.concurrent.ThreadPoolExecutor.runWorker + 1113(ThreadPoolExecutor.java:1113)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run + 588(ThreadPoolExecutor.java:588)
       at com.google.firebase.firestore.util.AsyncQueue$DelayedStartFactory.run + 205(AsyncQueue.java:205)
       at java.lang.Thread.run + 818(Thread.java:818)
```

Same devices and OSs (Android 5 and Android 6). Dependency tree shows only one place for Firestore SDK 20.2.0 and Proguard rules don't include it. I appreciate any help because this issue is affecting users of my app and is still #1 in Crashlytics.

Regards!

**Comment by schmidt-sebastian** [comment]
Do you mind sharing what `com.google.firebase.firestore.local.SQLiteSchema.ensureSequenceNumbers` looks like for you (if you open it in Android Studio). For me it looks like:

```
    private void ensureSequenceNumbers() {
        Query sequenceNumberQuery = new Query(this.db, "SELECT highest_listen_sequence_number FROM target_globals LIMIT 1");
        Long boxedSequenceNumber = (Long)sequenceNumberQuery.firstValue(SQLiteSchema$$Lambda$6.lambdaFactory$());
        Assert.hardAssert(boxedSequenceNumber != null, "Missing highest sequence number", new Object[0]);
        long sequenceNumber = boxedSequenceNumber;
        SQLiteStatement tagDocument = this.db.compileStatement("INSERT INTO target_documents (target_id, path, sequence_number) VALUES (0, ?, ?)");
        Query untaggedDocumentsQuery = (new Query(this.db, "SELECT RD.path FROM remote_documents AS RD WHERE NOT EXISTS (SELECT TD.path FROM target_documents AS TD WHERE RD.path = TD.path AND TD.target_id = 0) LIMIT ?")).binding(new Object[]{100});
        boolean[] resultsRemaining = new boolean[1];

        do {
            resultsRemaining[0] = false;
            untaggedDocumentsQuery.forEach(SQLiteSchema$$Lambda$7.lambdaFactory$(resultsRemaining, tagDocument, sequenceNumber));
        } while(resultsRemaining[0]);
    }
```

I don't see any references `com/google/firebase/firestore/L/Ga` and I wonder were those come from.

If you can reproduce this locally, does it still reproduce if you turn off ProGuard completely?

**Comment by rylexr** [comment]
Sure. This is how it looks like:

```
  /**
   * Ensures that each entry in the remote document cache has a corresponding sentinel row. Any
   * entries that lack a sentinel row are given one with the sequence number set to the highest
   * recorded sequence number from the target metadata.
   */
  private void ensureSequenceNumbers() {
    // Get the current highest sequence number
    SQLitePersistence.Query sequenceNumberQuery =
        new SQLitePersistence.Query(
            db, "SELECT highest_listen_sequence_number FROM target_globals LIMIT 1");
    Long boxedSequenceNumber = sequenceNumberQuery.firstValue(c -> c.getLong(0));
    hardAssert(boxedSequenceNumber != null, "Missing highest sequence number");

    long sequenceNumber = boxedSequenceNumber;
    SQLiteStatement tagDocument =
        db.compileStatement(
            "INSERT INTO target_documents (target_id, path, sequence_number) VALUES (0, ?, ?)");

    SQLitePersistence.Query untaggedDocumentsQuery =
        new SQLitePersistence.Query(
                db,
                "SELECT RD.path FROM remote_documents AS RD WHERE NOT EXISTS ("
                    + "SELECT TD.path FROM target_documents AS TD "
                    + "WHERE RD.path = TD.path AND TD.target_id = 0"
                    + ") LIMIT ?")
            .binding(SEQUENCE_NUMBER_BATCH_SIZE);

    boolean[] resultsRemaining = new boolean[1];

    do {
      resultsRemaining[0] = false;

      untaggedDocumentsQuery.forEach(
          row -> {
            resultsRemaining[0] = true;
            tagDocument.clearBindings();
            tagDocument.bindString(1, row.getString(0));
            tagDocument.bindLong(2, sequenceNumber);
            hardAssert(tagDocument.executeInsert() != -1, "Failed to insert a sentinel row");
          });
    } while (resultsRemaining[0]);
  }

```

I believe you're not seeing version 20.2.0. I checked library directly and code is the same as above. I can't reproduce it at all.

**Comment by rylexr** [comment]
BTW, line 364 (the one from stacktrace) is this:

`    Long boxedSequenceNumber = sequenceNumberQuery.firstValue(c -> c.getLong(0));
`

**Comment by schmidt-sebastian** [comment]
Did you resolve the source dependencies in Android Studio? I am more interested in seeing what the decompiled code looks like (which would not have any comments).  From your stacktrace it looks like `SQLiteSchema$$Lambda$6.lambdaFactory$()` is `com/google/firebase/firestore/L/Ga` in your project. I don't have an explanation for as to why though.

**Comment by rylexr** [comment]
What do you mean? I fired Gradle dependencies task and check Firestore SDK if that's what you are asking for. First time it was imported by 2 modules but now just by 1. Feel free to tell me what you require specifically and I'll gladly share it.

**Comment by schmidt-sebastian** [comment]
I was trying to get to the bottom of the binary dependencies that your app is pulling in. It is however extremely unlikely that your Maven dependencies differ from what we published officially, so it is probably not worth exploring this route any further. 

I'll see if I can come up with any other theories, but for now I am unfortunately drawing a blank.

**Comment by rylexr** [comment]
Not sure if this is relevant but I have jetifier enabled on my project:
```
android.useAndroidX=true
android.enableJetifier=true

```
I'm telling you this because Firestore SDK is under jetified-firebase-20.2.0 folder in my setup.

**Comment by rylexr** [comment]
Hi. Any news on this issue?

**Comment by schmidt-sebastian** [comment]
Sorry - I have not been able to reproduce this. Are you able to provide us with a small sample project? Does this also occur if you use our quickstart (https://github.com/firebase/quickstart-android/tree/master/firestore)? Thanks!

**Comment by TomislavMladenov** [comment]
Hello, I quit trying to figure it out. I just went to other database. Have no idea how can this happen on empty project as was mine. Maybe the problem was in me, don't know

**Comment by rylexr** [comment]
While I appreciate the help from you guys I don't understand this behavior of closing tickets automatically because "there are no new updates". There's a lot of information here provided by different users. Please help us fix it.

I managed to reproduce it on a HTC One M8 with Android 6. I still don't know why is happening but I believe it's related to AAB and Firebase on Android 5/6. If I install my application using Android Studio it works correctly. If however, I install the app through Play Store it fails with same error.

**Comment by cameronhenige** [comment]
@rylexr I am having pretty much the exact same issue in my app. https://stackoverflow.com/questions/57790871/how-to-fix-internal-error-in-firestore-crash. I am using AAB as well.

**Comment by schmidt-sebastian** [comment]
We will not be closing this issue as long as there is activity. We, however, still don't have a lead on what is causing these issues and have not been able to see this locally in our testing. I will see if I can track down someone internally that knows if AAB's packaging could explain this issue.

**Comment by cameronhenige** [comment]
Hey @schmidt-sebastian, I created a test app that reproduces this issue. https://github.com/cameronhenige/Reproduce-Crash. I ran the app on a Nexus 4 API 23 emulator to reproduce the crash. Please let me know if you need any more information from me. :)

**Comment by schmidt-sebastian** [comment]
@cameronhenige Thanks so much! It reproduces for me as well on the specified device (but not on others). I will try to see what is going on. We should not need more information.

**Comment by schmidt-sebastian** [comment]
@cameronhenige @rylexr @TomislavMladenov 

The minimum set of dependencies that reproduces this issue is:

```
dependencies {
    implementation 'androidx.appcompat:appcompat:1.0.2' // Needed for sample app
    implementation 'androidx.constraintlayout:constraintlayout:1.1.3' // Needed for sample app
    implementation 'com.google.firebase:firebase-firestore:21.0.0'
    implementation 'com.google.gms:google-services:X.Y.Z'  
}
```

You should be able to remove your implementation dependency on `com.google.gms:google-services`. `com.google.gms:google-services` should only be used at compile time as a plugin (`apply plugin 'com.google.gms.google-services'`), which allows it to inject code to configure Firebase from your GoogleService-Info.plist file. 

To apply `com.google.gms:google-services` as a plugin, you only have to add it to your classpath but not as a dependency (as done here: https://github.com/cameronhenige/Reproduce-Crash/blob/master/build.gradle#L18)

Furthermore, https://stackoverflow.com/questions/37360126/getting-exception-java-lang-noclassdeffounderror-com-google-firebase-firebaseop mentions that updating to Play Services Version 30 fixes this issue.

Please do let us know if this helps resolve your issue, in which case we can probably close this. We are sorry for the long time that it took us to find a fix/workaround.

**Comment by rylexr** [comment]
@schmidt-sebastian unfortunately that's the exact same configuration I have and still the problem is there. Here is how my config looks like:

**root level build.gradle**

```
buildscript {
    ext {
        fabric = '1.29.0'
        firebase_performance = '1.2.0'
        google_services = '4.3.0'
        gradle_tools = '3.6.0-alpha03'
        kotlin_version = '1.3.50'

        firebaseAnalytics = '17.2.0'
        firebaseAuth = '19.0.0'
        firebaseCrashlytics = '2.10.1'
        firebaseDatabase = '19.0.0'
        firebaseFirestore = '21.0.0'
        firebaseFunctions = '19.0.1'
        firebaseMessaging = '20.0.0'
        firebaseStorage = '19.0.0'
        firebasePerformance = '19.0.0'

        playCore = '1.6.1'
        playServicesAuth = '17.0.0'
        playServicesBase = '17.1.0'
        playServicesLocation = '16.0.0'
    }

    repositories {
        jcenter()
        google()
        maven { url 'https://maven.fabric.io/public' }
    }

    dependencies {
        classpath "com.android.tools.build:gradle:$gradle_tools"
        classpath "com.google.gms:google-services:$google_services"
        classpath "io.fabric.tools:gradle:$fabric"
        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
        // Performance Monitoring plugin
        classpath "com.google.firebase:firebase-plugins:$firebase_performance"
    }
}

```

**app level build.gradle**

```
apply plugin: 'com.android.application'
apply plugin: 'com.google.firebase.firebase-perf'
apply plugin: 'io.fabric'
apply plugin: 'kotlin-android'
apply plugin: 'kotlin-android-extensions'
apply plugin: 'kotlin-kapt'

repositories {
...
}
android {
    compileSdkVersion 28
    defaultConfig {
        minSdkVersion 21 // LP
        targetSdkVersion 28 // P
    }
    signingConfigs {
    ...
    }
   ...
}
dependencies {
...
}
apply plugin: 'com.google.gms.google-services'
```
Any ideas?

**Comment by cameronhenige** [comment]
@rylexr Sorry to hear you are still having issues. @schmidt-sebastian Thank you so much! You fixed the issue for me!

**Comment by schmidt-sebastian** [comment]
@rylexr Can you please share your app level dependencies?

```
dependencies {
...
}
```

**Comment by rylexr** [comment]
Sure. Here they are:

    debugImplementation "com.amitshekhar.android:debug-db:$debugDatabase"
    debugImplementation "com.squareup.leakcanary:leakcanary-android:$leakCanary"
    testImplementation
"com.squareup.leakcanary:leakcanary-android-no-op:$leakCanary"
    releaseImplementation
"com.squareup.leakcanary:leakcanary-android-no-op:$leakCanary"

    // *************
    // ** Testing **
    // *************
    // JUnit
    testImplementation "org.junit.jupiter:junit-jupiter-api:$junit"
    // Core library
    testImplementation "androidx.test:core-ktx:$androidTest"

    // ************
    // ** Google **
    // ************
    // Google Play Services
    implementation "com.google.android.play:core:$playCore"
    // Google Base Client (Proguard Exceptions are here)
    implementation "com.google.android.gms:play-services-base:$playServicesBase"
    // Google Sign-In
    implementation "com.google.android.gms:play-services-auth:$playServicesAuth"
    // Google Places
    implementation "com.google.android.libraries.places:places:$places"
    // Google Play Billing
    implementation "com.android.billingclient:billing:$billing"


    // *****************************
    // ** Architecture Components **
    // *****************************
    // Lifecycles
    implementation "androidx.lifecycle:lifecycle-extensions:$lifecycle"

    // Test helpers for LiveData
    testImplementation "androidx.arch.core:core-testing:$coreTesting"
    // Room
    implementation "androidx.room:room-runtime:$room"
    kapt "androidx.room:room-compiler:$room"
    // Room optional - Kotlin Extensions and Coroutines support for Room
    implementation "androidx.room:room-ktx:$room"
    // Test helpers for Room
    testImplementation "androidx.room:room-testing:$room"
    // * Paging *
    implementation "androidx.paging:paging-runtime:$paging"
    // Test helpers for Paging
    testImplementation "androidx.paging:paging-common:$paging"
    // * Navigation *
    implementation "android.arch.navigation:navigation-fragment:$navigation"
    implementation "android.arch.navigation:navigation-ui:$navigation"


    // **************
    // ** Firebase **
    // **************
    // Firebase Analytics
    implementation "com.google.firebase:firebase-core:$firebaseAnalytics"
    // Firebase Crash
    implementation
"com.crashlytics.sdk.android:crashlytics:$firebaseCrashlytics"
    // Firebase Auth
    implementation "com.google.firebase:firebase-auth:$firebaseAuth"
    // Firebase Database
    implementation "com.google.firebase:firebase-database:$firebaseDatabase"
    // Firebase Firestore
    implementation "com.google.firebase:firebase-firestore:$firebaseFirestore"
    // Firebase Functions
    implementation "com.google.firebase:firebase-functions:$firebaseFunctions"
    // Firebase: FCM
    implementation "com.google.firebase:firebase-messaging:$firebaseMessaging"
    // Firebase: Storage
    implementation "com.google.firebase:firebase-storage:$firebaseStorage"
    // Firebase: Performance
    implementation "com.google.firebase:firebase-perf:$firebasePerformance"


    // **************
    // ** Facebook **
    // **************
    // Facebook Login SDK
    implementation "com.facebook.android:facebook-login:$facebookLogin"

    // *************
    // ** Twitter **
    // *************
    // Twitter Login SDK
    implementation "com.twitter.sdk.android:twitter-core:$twitterLogin"
    implementation "com.twitter.sdk.android:twitter:$twitterLogin"


    // **********
    // ** Core **
    // **********
    // Guava
    api "com.google.guava:guava:$guava"
    // Google GSON
    implementation "com.google.code.gson:gson:$gson"
    // Kotlin
    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk8:$kotlin_version"
    // Kotlin Coroutines
    implementation
"org.jetbrains.kotlinx:kotlinx-coroutines-core:$kotlinCoroutines"
    implementation
"org.jetbrains.kotlinx:kotlinx-coroutines-android:$kotlinCoroutines"
    // Kotlin KTX Core
    implementation "androidx.core:core-ktx:$ktx"
    // EventBus
    implementation "org.greenrobot:eventbus:$eventbus"
    // Retrofit (REST client)
    implementation "com.squareup.retrofit2:retrofit:$retrofit"
    implementation "com.squareup.retrofit2:converter-gson:$retrofit"
    // Backport of JSR-310
    implementation "com.jakewharton.threetenabp:threetenabp:$threeTenABP"
    // Dagger
    implementation "com.google.dagger:dagger:$dagger"
    kapt "com.google.dagger:dagger-compiler:$dagger"
    // dagger.android
    implementation "com.google.dagger:dagger-android:$dagger"
    // if you use the support libraries
    implementation "com.google.dagger:dagger-android-support:$dagger"
    kapt "com.google.dagger:dagger-android-processor:$dagger"
    // Process Phoenix
    implementation "com.jakewharton:process-phoenix:$phoenix"


    // ********
    // ** UI **
    // ********
    // Fragment
    implementation "androidx.fragment:fragment-ktx:$fragment"
    // RecyclerView
    implementation "androidx.recyclerview:recyclerview:$recyclerview"
    // ConstraintLayout
    implementation
"androidx.constraintlayout:constraintlayout:$constraintLayout"
    // Material
    implementation "com.google.android.material:material:$material"
    // FlexBox Layout
    implementation "com.google.android:flexbox:$flexbox"
    // FastScroller
    implementation 'com.simplecityapps:recyclerview-fastscroll:2.0.0'
    // Glide
    implementation "com.github.bumptech.glide:glide:$glide"
    kapt "com.github.bumptech.glide:compiler:$glide"
    // Glide Transformations
    implementation "jp.wasabeef:glide-transformations:$glideTransformations"
    // ViewPager Indicator
    implementation
"fr.avianey.com.viewpagerindicator:library:$viewpagerIndicator"


----

        fabric = '1.29.0'
        firebase_performance = '1.2.0'
        google_services = '4.3.0'
        gradle_tools = '3.6.0-alpha03'
        kotlin_version = '1.3.50'

        androidTest = '1.2.1-alpha02'
        annotation = '1.1.0'
        arch = '1.1.1'

        billing = '2.0.0'

        cardstackview = '2.2.5'
        circleImageView = '3.0.0'
        circularReveal = '2.1.0'
        constraintLayout = '2.0.0-beta1'
        colorpicker = '0.0.13'
        coreTesting = '2.1.0-rc01'
        crashlytics = '2.10.1'

        dagger = '2.24'
        debugDatabase = '1.0.6'

        eventbus = '3.1.1'

        facebookLogin = '[5,6)'
        firebaseAnalytics = '17.2.0'
        firebaseAuth = '19.0.0'
        firebaseCrashlytics = '2.10.1'
        firebaseDatabase = '19.0.0'
        firebaseFirestore = '21.0.0'
        firebaseFunctions = '19.0.1'
        firebaseMessaging = '20.0.0'
        firebaseStorage = '19.0.0'
        firebasePerformance = '19.0.0'
        flexbox = '1.1.0'
        fragment = '1.2.0-alpha02'

        glide = '4.9.0'
        glideTransformations = '4.0.1'
        gson = '2.8.5'
        guava = '28.0-jre'

        junit = '5.5.1'

        kotlinCoroutines = '1.3.0'
        ktx = '1.2.0-alpha03'

        leakCanary = '1.6.3'
        lifecycle = '2.2.0-alpha03'

        material = '1.1.0-alpha09'
        mockito = '2.+'

        navigation = '1.0.0'

        pagerSlidingTabStrip = '1.1.1'
        paging = '2.1.0'
        places = '2.0.0'
        playCore = '1.6.1'
        playServicesAuth = '17.0.0'
        playServicesBase = '17.1.0'
        playServicesLocation = '16.0.0'
        phoenix = '2.0.0'

        recyclerview = '1.1.0-beta03'
        retrofit = '2.6.1'
        robolectric = '4.3'
        room = '2.2.0-beta01'

        threeTenABP = '1.2.1'
        twitterLogin = '3.3.0'

        viewpagerIndicator = '2.4.1.1@aar'

**Comment by schmidt-sebastian** [comment]
@rylexr Your crash is because of your dependency on Guava.

Note that if you change your guava version from '28.0-jre' to '28.0-android' this crash goes away.

For anyone else that is experiencing this issue: These NoClassDefFound errors seems to be introduced by pulling in dependencies that somehow conflict with Firestore. The way I debugged this is by disabling dependencies until the crash went away. I am not sure what the root source of these problems is, but it is likely not a problem with the Firestore SDK itself.

For @rylexr the crash manifested as:

```
Caused by: java.lang.ClassNotFoundException: Didn't find class "com.google.firebase.firestore.local.SQLiteMutationQueue$$Lambda$5" on path: DexPathList[[zip file "/data/app/com.example.reproducecrash-1/base.apk"],nativeLibraryDirectories=[/data/app/com.example.reproducecrash-1/lib/x86, /vendor/lib, /system/lib]]
        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:511)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:469)
```

Note that in the final APK, classes4.dex has the following:
       
<img width="298" alt="Screen Shot 2019-09-05 at 3 33 54 PM" src="https://user-images.githubusercontent.com/19397744/64388020-9926c180-cff2-11e9-84aa-428a57d4234b.png">

So the class is in fact included (as you can hopefully see despite my bad cropping).
